src/mlmc/quantity_concept.py

```diff
@@ -121,120 +373,594 @@ def mult_op(x, y):
     def __rmul__(self, other):
         if isinstance(other, (float, int)):
             return self.__const_mult(other)
-        assert(False)
+        assert False
 
     def __const_mult(self, other):
         def cmult_op(x, c=other):
             return c * x
         return self._reduction_op([self], cmult_op)
 
-def make_root_quantity(storage: Storage, q_specs : List[QSpec])
-    i_range = 0
-    dict_items = []
-    for q_spec in q_specs:
-        time_frames = []
-        for t in q_spec.times:
-            field_values = []
-            for loc in q_spec.locations:
-                q_range = QRange(storage)
-                array = Array.from_range(q_spec.shape, q_range)
-                q_range.set(i_range, array._size)
-                i_range += array._size
-                field_values.append( (loc, array) )
-            time_frames.append( (t, Field(field_values)) )
-        dict_items = (q_spec.name, TimeSeries(time_frames))
-    return Dict(dict_items)
-
-
-
-class QRange:
-    def __init__(self, storage):
-        self._storage = storage
+    @staticmethod
+    def _process_mask(x, y, operator, level_id):
+        """
+        Create samples mask
+        All values for sample must meet given condition, if any value doesn't meet the condition,
+        whole sample is eliminated
+        :param x: Quantity
+        :param y: Quantity or int, float
+        :param operator: operator module function
+        :param level_id: int, level identifier
+        :return: np.ndarray of bools
+        """
+        # Zero level - use just fine samples
+        if level_id == 0:
```

Comment:
Can we just reduce shape to (M,1,N) for level 0 starting from the storage and get rid of this special case?

Reply:
Make a separate issue as this change can possibly break another code.